marshal/marshal.go

```diff
@@ -315,62 +315,72 @@ func (m *Marshaler) ClaimPartition(topicName string, partID int) bool {
 // Heartbeat will send an update for other people to know that we're still alive and
 // still owning this partition. Returns an error if anything has gone wrong (at which
 // point we can no longer assert we have the lock).
-func (m *Marshaler) Heartbeat(topicName string, partID int, CurrentOffset int64) error {
+func (m *Marshaler) Heartbeat(topicName string, partID int, offset int64) error {
 	topic, err := m.getClaimedPartitionState(topicName, partID)
 	if err != nil {
 		return err
 	}
 
+	// Attempt to commit offset, this is best-effort and we don't care if it fails
+	// since the canonical storage is in the heartbeat
+	if err := m.CommitOffsets(topicName, partID, offset); err != nil {
+		log.Warningf("[%s:%d] failed to commit offset during heartbeat: %s",
+			topicName, partID, err)
+	}
+
 	// All good, let's heartbeat
 	cl := &msgHeartbeat{
 		msgBase:       *m.msgBase(topicName, partID),
-		CurrentOffset: CurrentOffset,
+		CurrentOffset: offset,
 	}
 	_, err = m.cluster.producer.Produce(MarshalTopic, int32(topic.claimPartition),
 		&proto.Message{Value: []byte(cl.Encode())})
 	if err != nil {
+		log.Errorf("[%s:%d] failed to send heartbeat message to Kafka: %s",
+			topicName, partID, err)
 		return fmt.Errorf("Failed to produce heartbeat to Kafka: %s", err)
 	}
 
-	err = m.CommitOffsets(topicName, partID, CurrentOffset)
-	return err
+	return nil
 }
 
 // ReleasePartition will send an update for other people to know that we're done with
 // a partition. Returns an error if anything has gone wrong (at which
 // point we can no longer assert we have the lock).
-func (m *Marshaler) ReleasePartition(topicName string, partID int, CurrentOffset int64) error {
+func (m *Marshaler) ReleasePartition(topicName string, partID int, offset int64) error {
 	topic, err := m.getClaimedPartitionState(topicName, partID)
 	if err != nil {
 		return err
 	}
 
+	// Commit our offset first; if this fails, we can still try to release,
+	// but we should advise
```

Comment:
Correct, I'm not sure how to address that at this time though. It seems non-trivial to fix given the inherent design of the system. I'm comfortable saying that this failure case is acceptable, though.

I suppose the alternative is that we make it error out the release if the CommitOffset call fails although that brings its own set of issues -- that degrades the normal reliability of the system I suspect.